robot/packages/deferred_package_manager.go

```diff
@@ -0,0 +1,174 @@
+
+// NewDeferredPackageManager returns a package manager that wraps a CloudPackageManager and a channel that is establishing
+// a connection to app. It starts up instantly and behaves as a noop package manager until a connection to app has been established.
+func NewDeferredPackageManager(
+	connectionChan chan DeferredConnectionResponse,
+	cloudConfig *config.Cloud,
+	packagesDir string,
+	logger logging.Logger,
+) ManagerSyncer {
+	noopManager := noopManager{Named: InternalServiceName.AsNamed()}
+	return &deferredPackageManager{
+		Named:             DeferredServiceName.AsNamed(),
+		logger:            logger,
+		cloudManagerArgs:  cloudManagerConstructorArgs{cloudConfig, packagesDir, logger},
+		connectionChan:    connectionChan,
+		noopManager:       noopManager,
+		lastSyncedManager: &noopManager,
+	}
+}
+
+// getCloudManager is the only function allowed to set or read m.cloudManager
+// every other function should use getCloudManager().
+func (m *deferredPackageManager) getCloudManager(wait bool) ManagerSyncer {
+	m.cloudManagerLock.Lock()
+	defer m.cloudManagerLock.Unlock()
+	// early exit if the cloud manager already exists
+	if m.cloudManager != nil {
+		return m.cloudManager
+	}
+	var response *DeferredConnectionResponse
+	if wait {
+		m.logger.Info("waiting for connection to app")
+		res := <-m.connectionChan
+		response = &res
+		m.logger.Info("connection to app established")
+	} else {
+		select {
+		case res := <-m.connectionChan:
+			response = &res
+		default:
+		}
+	}
+	if response == nil {
+		return nil
+	}
+	// because we have pulled this failed result from the connectionChan,
+	// it will retry
+	if response.Err != nil {
+		m.logger.Warnf("failed to establish a connection to app.viam %w", response.Err)
+		return nil
+	}
+	var err error
+	m.cloudManager, err = NewCloudManager(
+		m.cloudManagerArgs.cloudConfig,
+		response.Client,
+		m.cloudManagerArgs.packagesDir,
+		m.cloudManagerArgs.logger,
+	)
+	if err != nil {
+		m.logger.Warnf("failed to create cloud package manager %w", response.Err)
+	}
+	return m.cloudManager
+}
+
+// isMissingPackages is used pre-sync to determine if we should force-wait for the connection
+// to be established.
+func (m *deferredPackageManager) isMissingPackages(packages []config.PackageConfig) bool {
+	for _, pkg := range packages {
+		dir := pkg.LocalDataDirectory(m.cloudManagerArgs.packagesDir)
+		if _, err := os.Stat(dir); err != nil {
+			return true
+		}
+	}
+	return false
+}
+
+// Sync syncs all given packages and removes any not in the list from the local file system.
+// If there are packages missing on the local fs, this will wait while attempting to establish a connection to app.viam.
+func (m *deferredPackageManager) Sync(ctx context.Context, packages []config.PackageConfig) error {
+	shouldWait := m.isMissingPackages(packages)
+	var mgrToSyncWith ManagerSyncer
+	if mgr := m.getCloudManager(shouldWait); mgr != nil {
```

Comment:
architecture question -- there's nothing continuously waiting on connectionChan? the idea is that when I call Sync, it calls getCloudManager, which will block if m.cloudManager != nil

as soon as the goroutine in local_robot.go gives me a DeferredConnectionResponse, will it connect a second one which it holds on to? when I get disconnected, do I set m.cloudManager to nil so that it gets the waiting connection?

does this design double the # of package manager connections?

if my current connection fails, let's say because I lost wifi, will the waiting connection on the other side of the channel also fail? i.e. will I go through an extra retry loop?

if this is only used when Sync is called, why have the goroutine waiting to acquire?